### PR TITLE
Error handling cleanup in solvers/floatbv

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9df60c6d55b6bea494c57af0fed04bad37f1f1a3).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84700185

File: src/solvers/floatbv/float_utils.cpp [hunk 1]
```diff
@@ -84,58 +83,57 @@ bvt float_utilst::to_integer(
...
+    result.size() == dest_width,
+    "result bitvector width should equal the destination bitvector width");
+
+  // if signed, apply sign.
+  if(is_signed)
+    result = bv_utils.cond_negate(result, unpacked.sign);
   else
-    throw "unsupported rounding mode";
+  {
+    // It's unclear what the behaviour for negative floats
```
Comment (peterschrammel) [inline src/solvers/floatbv/float_utils.cpp:132]: `UNIMPLEMENTED;` ?

File: src/solvers/floatbv/float_bv.cpp [hunk 2]
```diff
@@ -810,9 +812,9 @@ exprt float_bvt::relation(
       not_exprt(nan));
   }
   else
-    assert(0);
+    UNREACHABLE;
```
Comment (tautschnig) [inline src/solvers/floatbv/float_bv.cpp]: Maybe this `else` branch should just be removed? It's redundant with the code afterwards.

File: src/solvers/floatbv/float_utils.cpp [hunk 3]
```diff
@@ -646,15 +644,16 @@ literalt float_utilst::relation(
       !NaN);
   }
   else
-    assert(0);
+    UNREACHABLE;
```
Comment (tautschnig) [inline src/solvers/floatbv/float_utils.cpp]: Redundant with the code afterwards, just remove the `else` branch

File: src/solvers/floatbv/float_utils.cpp (hunk 1 above)
Comment (danpoe) [reply src/solvers/floatbv/float_utils.cpp:132]: I think that's deliberate. According to the C standard, converting a negative float to an unsigned int is undefined behaviour. The code here chooses to simply ignore the sign bit when converting to an unsigned int (i.e., positive floats and their negative counterparts yield the same unsigned int value).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 98538b4fd4bb7f59e0c01cd827233317ec435de5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85736377

File: src/solvers/floatbv/float_utils.cpp (hunk 1 above)
Comment (peterschrammel) [reply src/solvers/floatbv/float_utils.cpp:132]: It seems this case is covered by `--conversion-check`. 
Wouldn't it be better for a verifier to make the result of the conversion nondeterministic? (That question goes probably beyond this PR, though).

Comment (danpoe) [reply src/solvers/floatbv/float_utils.cpp:132]: I just tried this and `--conversion-check` indeed finds this error. Yes, to be more sound for when `--conversion-check` is not enabled we could make the conversion result nondet but I'd say we leave this as future work for now.

File: src/solvers/floatbv/float_utils.cpp [hunk 4]
```diff
@@ -1158,9 +1162,6 @@ float_utilst::biased_floatt float_utilst::bias(const unbiased_floatt &src)
   // we need to bias the new exponent
   result.exponent=add_bias(src.exponent);
 
-  // strip off hidden bit
-  assert(src.fraction.size()==spec.f+1);
```
Comment (kroening) [inline src/solvers/floatbv/float_utils.cpp:1162]: I would have retained the comment.

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0dce17928aa9a82ee60af13763a5704f60a3cd7d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86879780

**Pull request merged**